### PR TITLE
Migrate DotProductAttention to NNX

### PR DESCRIPTION
**Pull request by hsuan-lun-chiang**
# Description

Migrate DotProductAttention to NNX.

- Use nnx_wrappers.ToNNX to bridge DotProductAttention, which imported from NVIDIA's transformer library, to NNX.

# Tests

Train Gemma-2b with attention=cudnn_flash_te, which use DotProductAttention:
```
python3 -m MaxText.train  MaxText/configs/base.yml run_name=gpt3-train-run base_output_directory=gs://maxtext-test/gemma-train/18/ max_target_length=128 model_name=gemma-2b dataset_type=synthetic steps=10 hardware=gpu attention=cudnn_flash_te
```

[Logs - After Migration](https://paste.googleplex.com/4861347309027328)
[Logs - Before Migration](https://paste.googleplex.com/6460959694258176)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/attention_op.py [hunk 1]
```diff
@@ -485,28 +485,76 @@ def maybe_create_nnx(einsum, *args):
...
+          transpose_batch_sequence=False,
+          window_size=sliding_window_size,
+          context_parallel_causal_load_balanced=config.context_parallel_load_balance,
+          context_parallel_axis="context",
+      )
+
+      dpa_layer = nnx_wrappers.ToNNX(dpa_layer, rngs=rngs)
+      dummy_query_prefill = jnp.zeros((1, self.max_target_length, self.num_query_heads, config.head_dim), dtype=self.dtype)
+      dummy_key_prefill = jnp.zeros((1, self.max_target_length, self.num_kv_heads, config.head_dim), dtype=self.dtype)
+      dummy_value_prefill = jnp.zeros((1, self.max_target_length, self.num_kv_heads, config.head_dim), dtype=self.dtype)
```
Comment (bvandermoon) [inline MaxText/layers/attention_op.py]: @cgarciae are zeros the right value here?

**Comment by hsuan-lun-chiang** [comment]
> Thank you @hsuan-lun-chiang. Could you check why the GPU integration test is failing?

Sure! It was causes by None being casting to uint8, fixed it. Thank you.

File: MaxText/layers/attention_op.py [hunk 2]
```diff
@@ -1096,48 +1145,23 @@ def cudnn_flash_attention(
...
-        qkv_layout="BSHD_BSHD_BSHD",  # 'BS3HD', 'BSHD_BS2HD' or 'BSHD_BSHD_BSHD'
-        scale_factor=1.0,
-        transpose_batch_sequence=False,
-        window_size=sliding_window_size,
-        context_parallel_causal_load_balanced=self.config.context_parallel_load_balance,
-        context_parallel_axis="context",
-    )
-    return dpa_layer(query, key, value, mask=attn_mask)
+    if attn_mask is not None:
+      attn_mask = jnp.where((attn_mask >= DEFAULT_MASK_VALUE * 0.5), 0, 1)
```
Comment (hsuan-lun-chiang) [inline MaxText/layers/attention_op.py]: Hi @bvandermoon, I hit a Type AssertionError when training gemma-2b with cudnn_flash_te in **main branch**. The stack trace is [here](https://paste.googleplex.com/6258226257199104).

It looks like the mask's data type was the issue. I've added a cast to uint8 based on [this](https://github.com/AI-Hypercomputer/maxtext/blob/main/MaxText/layers/attention_op.py#L109).

Would you mind sharing your thoughts on this fix? I'd appreciate your feedback.

**Review by bvandermoon — APPROVED** [review]
LGTM. Thank you @hsuan-lun-chiang. Did you run the description test on a GPU VM? Let's get @cgarciae's thoughts as well

**Comment by hsuan-lun-chiang** [comment]
> LGTM. Thank you @hsuan-lun-chiang. Did you run the description test on a GPU VM? Let's get @cgarciae's thoughts as well

Happy to help! Yes, I ran the test on a GPU VM with A100 80GB.

File: src/MaxText/layers/attention_op.py [hunk 3]
```diff
@@ -523,6 +523,55 @@ def maybe_create_nnx(einsum, *args):
       self.AqtEinsum_2 = jnp.einsum
       self.AqtEinsum_3 = jnp.einsum
 
+    if self.attention_kernel == "cudnn_flash_te":
```
Comment (RissyRan) [inline src/MaxText/layers/attention_op.py]: Wait, this is flash attention on GPU, not dot product.

https://github.com/AI-Hypercomputer/maxtext/blob/4a261bf818d779f0d3661daf8d505ab5b3d9ab80/src/MaxText/layers/attention_op.py#L911-L922

Is this change on purpose? cc @bvandermoon

Comment (hsuan-lun-chiang) [reply src/MaxText/layers/attention_op.py]: Thank you @RissyRan for pointing this out. After reviewing it again, I agree—this shouldn't be specific to FlashAttention. I've updated the code accordingly.

Comment (bvandermoon) [reply src/MaxText/layers/attention_op.py]: @hsuan-lun-chiang could you please comment on this?

Comment (hsuan-lun-chiang) [reply src/MaxText/layers/attention_op.py]: Hi @bvandermoon,
In my previous implementation, the code checked if the attention type was `cudnn_flash_attention` in the `AttentionOp.__init__`, and based on that, it instantiated a `DotProductAttention` (dpa_layer) and called ToNNX to handle the conversion.

However, upon reviewing it again, I realized that it's sufficient to convert just the DotProductAttention inside cudnn_flash_attention. I've updated the code accordingly.

Please let me know if you have any concerns or suggestions. Thank you.

Comment (hsuan-lun-chiang) [reply src/MaxText/layers/attention_op.py]: Sorry, I didn't realize my comments were in pending status and therefore not visible.

File: src/MaxText/layers/attention_op.py [hunk 4]
```diff
@@ -467,6 +467,7 @@ def __init__(
     self.chunk_attn_window_size = chunk_attn_window_size
     self.use_ragged_attention = use_ragged_attention
     self.ragged_block_size = ragged_block_size
+    self.rngs = rngs
```
Comment (cgarciae) [inline src/MaxText/layers/attention_op.py]: Fork to avoid sharing `rngs`.
```suggestion
    self.rngs = rngs.fork()
```

**Review by cgarciae — APPROVED** [review]

File: src/MaxText/layers/attention_op.py (hunk 4 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/attention_op.py]: Done. Thank you.

Comment (hsuan-lun-chiang) [reply src/MaxText/layers/attention_op.py]: Hi @cgarciae,
I found that using `fork()` causes google3 internal test `compare_output_and_grad_equivalence` to fail because the output values of `DeepSeekMoELayer` in `deepseek_batchsplit` and `deepseek` do not match.

Comment (hsuan-lun-chiang) [reply src/MaxText/layers/attention_op.py]: Hi @cgarciae,
I may have to undo this because fork() causes google3 internal test compare_output_and_grad_equivalence to fail - the output values of DeepSeekMoELayer in deepseek_batchsplit and deepseek do not match.
Please let me know if I misunderstood anything, thank you!
